hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/PulsarSource.java

```diff
@@ -0,0 +1,292 @@
+    } catch (PulsarClientException e) {
+      LOG.error(String.format("Failed to ack messageId (%s) for topic '%s'", latestConsumedOffset, topicName), e);
+      throw new HoodieIOException("Failed to ack message for topic", e);
+    }
+  }
+
+  private MessageId fetchLatestOffset() {
+    try {
+      return pulsarConsumer.get().getLastMessageId();
+    } catch (PulsarClientException e) {
+      LOG.error(String.format("Failed to fetch latest messageId for topic '%s'", topicName), e);
+      throw new HoodieIOException("Failed to fetch latest messageId for topic", e);
+    }
+  }
+
+  private Consumer<byte[]> subscribeToTopic() {
+    try {
+      // NOTE: We're generating unique subscription-id to make sure that subsequent invocation
+      //       of the DS, do not interfere w/ each other
+      String subscriptionId = String.format(HUDI_PULSAR_CONSUMER_ID_FORMAT, System.currentTimeMillis());
+      return pulsarClient.get()
+          .newConsumer()
+          .topic(topicName)
+          .subscriptionName(subscriptionId)
+          .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+          .subscriptionType(SubscriptionType.Exclusive)
+          .subscribe();
+    } catch (PulsarClientException e) {
+      LOG.error(String.format("Failed to subscribe to Pulsar topic '%s'", topicName), e);
+      throw new HoodieIOException("Failed to subscribe to Pulsar topic", e);
+    }
+  }
+
+  private PulsarClient initPulsarClient() {
+    try {
+      return PulsarClient.builder()
+          .serviceUrl(serviceEndpointURL)
+          .build();
+    } catch (PulsarClientException e) {
+      LOG.error(String.format("Failed to init Pulsar client connecting to '%s'", serviceEndpointURL), e);
+      throw new HoodieIOException("Failed to init Pulsar client", e);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    shutdownPulsarClient(pulsarClient.get());
+  }
+
+  private static Long computeTargetRecordLimit(long sourceLimit, TypedProperties props) {
+    if (sourceLimit < Long.MAX_VALUE) {
+      return sourceLimit;
+    } else {
+      return props.getLong(Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.key(),
+          Config.PULSAR_SOURCE_MAX_RECORDS_PER_BATCH_THRESHOLD_PROP.defaultValue());
+    }
+  }
+
+  private static String convertToOffsetString(String topic, MessageId startingOffset) {
+    return JsonUtils.topicOffsets(
+        HoodieConversionUtils.mapAsScalaImmutableMap(
+            Collections.singletonMap(topic, startingOffset)));
+  }
+
+  private static void shutdownPulsarClient(PulsarClient client) throws PulsarClientException {
+    client.close();
+    // NOTE: Current version of Pulsar's client (in Pulsar Spark Connector 3.1.1.4) is not
+    //       shutting down event-loop group properly, so we had to shut it down manually
+    try {
+      EventLoopGroup eventLoopGroup = ((PulsarClientImpl) client).eventLoopGroup();
```

Comment:
after calling close() in L 240, will the eventLoopGroup will still be accessible w/ the client instance?